File: forest.cpp
#include <algorithm>
#include <functional>
#include <iostream>
#include <vector>

using namespace std;

struct Forest {
  Forest(int g, int w, int l) : goats{g}, wolves{w}, lions{l} {}
  bool operator<(const Forest& rhs) const {
    if (goats != rhs.goats) return goats < rhs.goats;
    if (wolves != rhs.wolves) return wolves < rhs.wolves;
    return lions < rhs.lions;
  }
  bool operator==(const Forest& rhs) const {
    return (goats == rhs.goats) && (wolves == rhs.wolves) && (lions == rhs.lions);
  }
  int goats;
  int wolves;
  int lions;
};

bool is_stable(const Forest &f) {
  return (!f.goats && (!f.wolves || !f.lions)) || (!f.wolves && !f.lions);
}

bool is_invalid(const Forest &f) {
  return (f.goats < 0 || f.wolves < 0 || f.lions < 0);
}

ostream &operator<<(ostream &os, const Forest &f) {
  return os << "{ goats: " << f.goats
            << ", wolves: " << f.wolves
            << ", lions: " << f.lions << " }";
}

vector<Forest> mutate(const vector<Forest> &curr) {
  vector<Forest> next;
  next.reserve(curr.size() * 3);

  for (auto f : curr) {
    next.emplace_back(f.goats - 1, f.wolves - 1, f.lions + 1);
    next.emplace_back(f.goats - 1, f.wolves + 1, f.lions - 1);
    next.emplace_back(f.goats + 1, f.wolves - 1, f.lions - 1);
  }

  // filter out invalid forests
  auto valid_end = remove_if(begin(next), end(next), is_invalid);
  // delete duplicates
  stable_sort(begin(next), valid_end);
  next.erase(unique(begin(next), valid_end), end(next));
  return next;
}

vector<Forest> solve(const Forest &x) {
  vector<Forest> xs = { x };
  while (!xs.empty() && none_of(begin(xs), end(xs), is_stable)) {
    xs = mutate(xs);
  }
  vector<Forest> stable;
  copy_if(begin(xs), end(xs), back_inserter(stable), is_stable);
  return stable;
}

int main(int argc, char *argv[]) {
  if (argc != 4) {
    cerr << "USAGE: " << argv[0] << " <goats> <wolves> <lions>" << endl;
    exit(EXIT_FAILURE);
  }
  Forest initial{stoi(argv[1]), stoi(argv[2]), stoi(argv[3])};
  for (auto f: solve(initial)) cout << "Solution: " << f << endl;
}

File: forest.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

type Forest struct {
	goats  int32
	wolves int32
	lions  int32
}

func (f Forest) IsStable() bool {
	if f.goats == 0 {
		return f.wolves == 0 || f.lions == 0
	}
	return f.wolves == 0 && f.lions == 0
}

func (f Forest) IsValid() bool {
	return f.goats >= 0 && f.wolves >= 0 && f.lions >= 0
}

type ForestSet map[Forest]struct{}

func (fs ForestSet) AnyStable() bool {
	for forest := range fs {
		if forest.IsStable() {
			return true
		}
	}
	return false
}

func mutate(forestSet ForestSet) ForestSet {
	potentials := make([]Forest, 0, len(forestSet)*3)
	for forest := range forestSet {
		potentials = append(potentials,
			Forest{forest.goats - 1, forest.wolves - 1, forest.lions + 1},
			Forest{forest.goats - 1, forest.wolves + 1, forest.lions - 1},
			Forest{forest.goats + 1, forest.wolves - 1, forest.lions - 1},
		)
	}

	newForestSet := make(ForestSet, len(potentials))
	for _, forest := range potentials {
		if forest.IsValid() {
			newForestSet[forest] = struct{}{}
		}
	}
	return newForestSet
}

func solve(initialForest Forest) []Forest {
	forestSet := ForestSet{initialForest: struct{}{}}
	for len(forestSet) > 0 && !forestSet.AnyStable() {
		forestSet = mutate(forestSet)
	}
	// return the stable remains
	stable := make([]Forest, 0, len(forestSet))
	for forest := range forestSet {
		if forest.IsStable() {
			stable = append(stable, forest)
		}
	}
	return stable
}

func main() {
	if len(os.Args) != 4 {
		fmt.Println("USAGE: forest <goats> <wolves> <lions>")
		os.Exit(1)
	}
	goats, _ := strconv.Atoi(os.Args[1])
	wolves, _ := strconv.Atoi(os.Args[2])
	lions, _ := strconv.Atoi(os.Args[3])

	initial := Forest{int32(goats), int32(wolves), int32(lions)}
	fmt.Printf("Initial: %v\n", initial)

	for _, f := range solve(initial) {
		fmt.Printf("Solution: %v\n", f)
	}
}
